exec/bigmachine.go

```diff
@@ -1275,54 +1256,132 @@ func (r *retryReader) Close() error {
+	return r.readCloser.Close()
+}
+
+// machineTaskPartition is a task partition on a specific machine. It implements
+// the openerAt interface to provide an io.ReadCloser to read the task data.
+type machineTaskPartition struct {
 	// Machine is the machine from which task data is read.
 	Machine *bigmachine.Machine
 	// TaskPartition is the task and partition that should be read.
 	TaskPartition taskPartition
+}
 
-	reader sliceio.Reader
-	rpc    *retryReader
+// OpenAt implements openerAt.
+func (m machineTaskPartition) OpenAt(ctx context.Context, offset int64) (reader io.ReadCloser, err error) {
+	err = m.Machine.RetryCall(ctx, "Worker.Read",
+		readRequest{m.TaskPartition.Name, m.TaskPartition.Partition, offset}, &reader)
+	return
 }
 
-func newMachineReader(machine *bigmachine.Machine, partition taskPartition) *machineReader {
-	m := &machineReader{
-		Machine:       machine,
-		TaskPartition: partition,
+func (m machineTaskPartition) String() string {
+	return fmt.Sprintf("Worker.Read %s:%s:%d", m.Machine.Addr, m.TaskPartition.Name, m.TaskPartition.Partition)
+}
+
+// newMachineReader returns a reader that reads a taskPartition from a machine.
+// It issues the (streaming) read RPC on the first call to Read so that data
+// are not buffered unnecessarily.
+func newMachineReader(machine *bigmachine.Machine, taskPartition taskPartition) *openerAtReader {
+	return &openerAtReader{
+		OpenerAt: machineTaskPartition{
+			Machine:       machine,
+			TaskPartition: taskPartition,
+		},
+		// This is how all slice operations read data to process and does not
+		// involve application code. By revising the severity, we save slice
+		// operation implementations from each individually revising the
+		// severity of machine reads.
+		ReviseSeverity: true,
 	}
-	return m
 }
 
-func (m *machineReader) Read(ctx context.Context, f frame.Frame) (int, error) {
-	if m.rpc == nil {
-		name := fmt.Sprintf("Worker.Read %s:%s:%d",
-			m.Machine.Addr, m.TaskPartition.Name, m.TaskPartition.Partition)
-		openerAt := machineTaskPartition{
-			machine:       m.Machine,
-			taskPartition: m.TaskPartition,
-		}
-		m.rpc = newRetryReader(ctx, name, openerAt)
-		m.reader = sliceio.NewDecodingReader(m.rpc)
+// evalOpenerAt is an openerAt that opens a reader for a task partition, first
+// evaluating the task to ensure that it is available. It is used for fault
+// tolerance of post-evaluation reads.
+type evalOpenerAt struct {
+	// Executor is the executor used to execute the task before opening the
+	// reader.
+	Executor *bigmachineExecutor
+	// Task is the task whose data is read by the returned reader.
+	Task *Task
+	// Partition is the data partition read by the returned reader.
+	Partition int
+
+	// machine is the machine used by the last attempt to open a reader,
+	// post-successful evaluation.
+	machine *bigmachine.Machine
+}
+
+// OpenAt implements openerAt.
+func (r *evalOpenerAt) OpenAt(ctx context.Context, offset int64) (reader io.ReadCloser, err error) {
```

Comment:
I think these return values should not be named: https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters.